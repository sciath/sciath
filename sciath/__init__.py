""" SciATH: Scientific Application Test Harness """
from sciath._sciath_io import NamedColors

__version__ = (0, 7, 0)

# A default set of colors
SCIATH_COLORS = NamedColors()
