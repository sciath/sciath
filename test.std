# this is a test file

$verify_str abcdf__efsss

  $verify = { 2.2 }


$verify { 
3.33,
6.6 }

KSP its 4

$verify = 99.9

$verify 51.0
