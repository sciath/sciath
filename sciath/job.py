""" SciATH Job class """
from __future__ import print_function

import sciath.task


class Job:
    r""":class:`Job` describes a ordered list of :class:`Task`\s.

    It describes the :class:`Task`\s themselves, not information about any
    particular "run" (handled within :class:`Harness`) or how one might
    interpret the results of such a run (handled by :class:`Test`).  A
    :class:`Launcher` object executes the :class:`Task`\s described by a :class:`Job`.

    Data include

    * A name
    * An ordered list of :class:`Task`\s

    It also defines how the name may be converted to various standard filenames.

    """

    _default_job_name = 'job'

    def __init__(self, task_or_tasks, name=None):
        if name is None:
            self.name = Job._default_job_name
            self.named_by_default = True
        else:
            self.name = name.replace(' ', '_')
            self.named_by_default = False

        if isinstance(task_or_tasks, sciath.task.Task):
            self.tasks = [task_or_tasks]
        elif isinstance(task_or_tasks, list):
            for task in task_or_tasks:
                if not isinstance(task, sciath.task.Task):
                    raise Exception('A Task or list of Tasks was expected')
                self.tasks = task_or_tasks
        else:
            raise Exception('A Task or list of Tasks was expected')

    def create_execute_command(self):
        """
        Returns a list containing (command, resource) tuples for the Job.
        """
        return [task.create_execute_command() for task in self.tasks]

    @property
    def complete_filename(self):
        """ Returns a filename for a sentinel signifying completion """
        return '.%s.complete' % self.name

    @property
    def exitcode_filename(self):
        """ Returns a filename to use for exit codes """
        return self.name + '.exitcode'

    @property
    def launched_filename(self):
        """ Returns a filename for a sentinel signifying launching """
        return '.%s.launched' % self.name

    @property
    def stdout_filename(self):
        """ Returns a filename to use for stdout """
        return self.name + '.stdout'

    @property
    def stderr_filename(self):
        """ Returns a filename to use for stderr """
        return self.name + '.stderr'

    # This can be removed once the transition to using templates is complete
    def get_max_resources(self):
        """ Returns a dict() defining the maximum required counts / values """

        # Use the first Task to determine which resources to consider
        max_resources = {}
        for key, value in self.tasks[0].resources.items():
            max_resources[key] = value

        for task in self.tasks[1:]:
            for key, value in task.resources.items():
                if value > max_resources[key]:
                    max_resources[key] = value
        return max_resources

    def resource_max(self, key):
        """ Returns the maximum value of a resource over all Tasks

            Returns None if no Tasks define that resource.
        """
        maximum = None
        for task in self.tasks:
            if key in task.resources:
                current = task.resources[key]
                if maximum is None:
                    maximum = current
                else:
                    maximum = max(maximum, current)
        return maximum

    def number_tasks(self):
        """ Returns the number of tasks within the Job """
        return len(self.tasks)

    def total_wall_time(self):
        """ Returns the total wall time required for all Tasks

            Returns None if any task does not have a defined wall time.
        """
        total_wall_time = 0
        for task in self.tasks:
            if task.wall_time is None:
                return None
            total_wall_time += task.wall_time
        return total_wall_time
