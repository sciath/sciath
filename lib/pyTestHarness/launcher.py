from __future__ import print_function
import os
import sys
import shutil
import pyTestHarness.test
from   pyTestHarness.colors import NamedColors as pthcolors
from   pyTestHarness.version import getVersion
from   pyTestHarness.utils import py23input

class PthTestHarnessLoadException(Exception) :
  pass

def FormattedHourMin(seconds):
  m, s = divmod(int(seconds), 60)
  h, m = divmod(m, 60)
  wt = "%02d:%02d" % (h, m)
  return(wt)

def FormattedHourMinSec(seconds):
  m, s = divmod(seconds, 60)
  h, m = divmod(m, 60)
  wt = "%02d:%02d:%02d" % (h, m,s)
  return(wt)

def pthFormatMPILaunchCommand(mpiLaunch,ranks,corespernode):
  launch = mpiLaunch
  launch = launch.replace("<ranks>",str(ranks))
  launch = launch.replace("<cores>",str(ranks))
  launch = launch.replace("<tasks>",str(ranks))
  launch = launch.replace("<RANKS>",str(ranks))
  if corespernode:
    launch = launch.replace("<corespernode>",str(corespernode))
    launch = launch.replace("<cores_per_node>",str(corespernode))
    launch = launch.replace("<CORESPERNODE>",str(corespernode))
    launch = launch.replace("<CORES_PER_NODE>",str(corespernode))
    launch = launch.replace("<rankspernode>",str(corespernode))
    launch = launch.replace("<ranks_per_node>",str(corespernode))
    launch = launch.replace("<RANKSPERNODE>",str(corespernode))
    launch = launch.replace("<RANKS_PER_NODE>",str(corespernode))
  return(launch)

def generateLaunch_PBS(accountname,queuename,testname,mpiLaunch,executable,ranks,ranks_per_node,walltime,outfile):
  if not ranks:
    print("<generateLaunch_PBS>: Requires the number of MPI-ranks be specified")
  if not walltime:
    print("<generateLaunch_PBS>: Requires the walltime be specified")

  filename = testname + '-pth.pbs'
  file = open(filename,"w")
  file.write("#!/bin/bash\n")

  file.write("# pth: auto-generated pbs file\n")

  if accountname:
    file.write("#PBS -A " + accountname + "\n") # account to charge
  file.write("#PBS -N \"" + testname + "\"" + "\n") # jobname
  file.write("#PBS -o " + testname + ".stdout" + "\n")
  file.write("#PBS -e " + testname + ".stderr" + "\n")

  if queuename:
    file.write("#PBS -q " + queuename + "\n")

  wt = FormattedHourMinSec(walltime*60.0)
  file.write("#PBS -l mppwidth=1024,walltime=" + wt + "\n")

  launch = pthFormatMPILaunchCommand(mpiLaunch,ranks,ranks_per_node)
  file.write(launch + " " + executable + " > " + outfile + "\n\n") # launch command
  file.close()
  return(filename)

def generateLaunch_SLURM(accountname,queuename,testname,constraint,mpiLaunch,execute,ranks,ranks_per_node,walltime,outfile):
  if not ranks:
    print("<generateLaunch_SLURM>: Requires the number of MPI-ranks be specified")
  if not walltime:
    print("<generateLaunch_SLURM>: Requires the walltime be specified")

  filename = testname + '-pth.slurm'
  file = open(filename,"w")
  file.write("#!/bin/bash -l\n")

  file.write("# pth: auto-generated slurm file\n")
  if accountname:
    file.write("#SBATCH --account=" + accountname + "\n") # account to charge
  file.write("#SBATCH --job-name=\"" + testname + "\"" + "\n") # jobname

  file.write("#SBATCH --output=" + testname + ".stdout" + "\n") # jobname.stdout
  file.write("#SBATCH --error=" + testname + ".stderr" + "\n") # jobname.stderr

  if queuename:
    file.write("#SBATCH --partition=" + queuename + "\n")

  file.write("#SBATCH --ntasks=" + str(ranks) + "\n")
  if ranks_per_node:
    file.write("#SBATCH --ntasks-per-node=" + str(ranks_per_node) + "\n")

  wt = FormattedHourMinSec(walltime*60.0)
  file.write("#SBATCH --time=" + wt + "\n")

  if constraint :
    file.write("#SBATCH --constraint=" + constraint + "\n")

  launch = pthFormatMPILaunchCommand(mpiLaunch,ranks,ranks_per_node)
  for e in execute:
    file.write(launch + " " + e + " >> " + outfile + "\n") # launch command
  file.write("\n")
  file.close()
  return(filename)

def generateLaunch_LSF(accountname,queuename,testname,mpiLaunch,execute,ranks,rusage,walltime,outfile):
  if not ranks:
    print("<generateLaunch_LSF>: Requires the number of MPI-ranks be specified")
  if not walltime:
    print("<generateLaunch_LSF>: Requires the walltime be specified")

  filename = testname + '-pth.lsf'
  file = open(filename,"w")
  file.write("#!/bin/sh\n")

  file.write("# pth: auto-generated lsf file\n")

  file.write("#BSUB -J " + testname + "\n") # jobname

  file.write("#BSUB -o " + testname + ".stdout\n") # jobname.stdout
  file.write("#BSUB -e " + testname + ".stderr\n") # jobname.stderr

  if queuename:
    file.write("#BSUB -q " + queuename + "\n")

  file.write("#BSUB -n " + str(ranks) + "\n")

  if rusage:
    file.write("#BSUB -R \'" + rusage + "\'" + "\n")

  wt = FormattedHourMin(walltime*60.0) 
  file.write("#BSUB -W " + wt + "\n")

  launch = pthFormatMPILaunchCommand(mpiLaunch,ranks,None)
  for e in execute:
    file.write(launch + " " + e + " >> " + outfile + "\n") # launch command
  file.write("\n")
  file.close()
  return(filename)

def generateLaunch_LoadLevelerBG(accountname,queuename,testname,execute,total_ranks,machine_ranks_per_node,walltime):
  if not total_ranks:
    print("<generateLaunch_LoadLeveler>: Requires the number of MPI-ranks be specified")
  if not walltime:
    print("<generateLaunch_LoadLeveler>: Requires the walltime be specified")

  print("#!/bin/sh")
  print("# pth: auto-generated llq file")
  print("# @ job_name = " + testname)
  print("# @ job_type = bluegene")
  print("# @ error = $(job_name)_$(jobid).stderr")
  print("# @ output = $(job_name)_$(jobid).stdout")
  print("# @ environment = COPY_ALL;")
  wt = FormattedHourMinSec(walltime*60.0)
  print("# @ wall_clock_limit = " + wt)
  print("# @ notification = never")
  print("# @ class = " + queuename)

  bgsize = math.ceil(total_ranks/machine_ranks_per_node)
  print("# @ bg_size = " + str(bgsize))
  print("# @ bg_connection = TORUS")
  print("# @ queue")

  for e in execute:
    print("runjob -n " + e) # launch command


class Launcher:
  confFileName = 'pthBatchQueuingSystem.conf'

  @staticmethod
  def writeDefaultDefinition():
    file = open(Launcher.confFileName,'w')
    major,minor,patch=getVersion()
    file.write('majorVersion=' + str(major) + '\n')
    file.write('minorVersion=' + str(minor) + '\n')
    file.write('patchVersion=' + str(patch) + '\n')
    file.write('queuingSystemType=none\n' )
    file.write('mpiLaunch=none\n' )
    file.close()

  def __init__(self):
    self.accountName = []
    self.queueName = []
    self.mpiLaunch = []
    self.queuingSystemType = []
    self.batchConstraint = []
    self.jobSubmissionCommand = []
    self.useBatch = False
    self.verbosity_level = 1

    self.setup()

    if self.useBatch :
      if self.mpiLaunch == 'none':
        raise RuntimeError('[pth] If using a queuing system, a valid mpi launch command must be provided')

  def setVerbosityLevel(self,value):
    self.verbosity_level = value

  def setQueueName(self,name):
    self.queueName = name

  def setBatchConstraint(self,argstring):
    self.batchConstraint=argstring

  def setHPCAccountName(self,name):
    self.accountName = name

  def setMPILaunch(self,name):
    self.mpiLaunch = name
    # check for existence of "rank" keyword in the string "name"
    if self.queuingSystemType in ['none', 'None', 'local'] and name != 'none':
      keywordlist = [ '<ranks>', '<cores>', '<tasks>', '<RANKS>' ]
      # check of any of keywordlist[i] appears in name
      valid_launcher = False
      for kword in keywordlist:
        if kword in name:
          valid_launcher = True
          break

      if valid_launcher == False:
        raise RuntimeError('[pth] Your MPI launch command must contain the keyword \"<ranks>\"')

  def setQueueSystemType(self,type):
    if type in ['PBS','pbs']:
      self.queuingSystemType = 'pbs'
      self.jobSubmissionCommand = 'qsub '
      self.useBatch = True
      #print('Recognized PBS queuing system')

    elif type in ['LSF','lsf']:
      self.queuingSystemType = 'lsf'
      self.jobSubmissionCommand = 'bsub < '
      self.useBatch = True
      #print('Recognized LSF queuing system')

    elif type in ['SLURM','slurm']:
      self.queuingSystemType = 'slurm'
      self.jobSubmissionCommand = 'sbatch '
      self.useBatch = True
      #print('Recognized Slurm queuing system')

    elif type in ['LoadLeveler','load_leveler','loadleveler','llq']:
      self.queuingSystemType = 'load_leveler'
      self.jobSubmissionCommand = 'llsubmit '
      self.useBatch = True
      #print('Recognized IBM LoadLeveler queuing system')

    elif type in ['none', 'None', 'local']:
      self.queuingSystemType = 'none'
      self.jobSubmissionCommand = ''
      #print('No queuing system being used')

    else:
      print('Value found: ' + type + ' ...')
      raise RuntimeError('[pth] Unknown or unsupported batch queuing system specified')

  def setQueueName(self,name):
    self.queueName = name

  def view(self):
    print('pth: Batch queueing system configuration [',Launcher.confFileName,']')
    major,minor,patch=getVersion()
    print('  Version:         ',str(major)+'.'+str(minor)+'.'+str(patch))
    print('  Queue system:    ',self.queuingSystemType)
    print('  MPI launcher:    ',self.mpiLaunch)
    if self.useBatch:
      print('  Submit command:', self.jobSubmissionCommand)
      if self.accountName:
        print('  Account:       ',self.accountName)
      if self.queueName:
        print('  Queue:         ',self.queueName)
      if self.batchConstraint :
        print('  Constraint:    ', self.batchConstraint)

  def configure(self):
    print('----------------------------------------------------------------')
    print('Creating new',Launcher.confFileName,'file')
    prompt = '[1] Batch queuing system type <pbs,lsf,slurm,llq,none>: '
    v = py23input(prompt)
    if not v:
      raise ValueError('[pth] You must specify the type of queuing system')
    self.setQueueSystemType(v)

    v = None
    while not v:
      prompt = '[2] MPI launch command with num. procs. flag (required - hit enter for examples): '
      v = py23input(prompt)
      if not v :
        print(' Required. Some example MPI launch commands:')
        print('  No MPI Required           : none')
        print('  Local Machine (mpirun)    : mpirun -np <ranks>')
        print('  Local Machine (mpiexec)   : mpiexec -np <ranks>')
        print('  SLURM w/ aprun            : aprun -B')
        print('  Native SLURM              : srun -n $SLURM_NTASKS')
        print('  LSF (Euler)               : mpirun')
        PETSC_DIR=os.getenv('PETSC_DIR')
        PETSC_ARCH=os.getenv('PETSC_ARCH')
        if PETSC_DIR and PETSC_ARCH:
          print('  Current PETSc MPI wrapper :',os.path.join(PETSC_DIR,PETSC_ARCH,'bin','mpiexec'),'-n <ranks>')
        else :
          print('  Example PETSc MPI wrapper : /users/myname/petsc/arch-xxx/bin/mpiexec -n <ranks>')
        print(' Note that the string \"<ranks>\" must be included if the number of ranks is required at launch.')
        print(' The keyword <ranks> will be replaced by the actual number of MPI ranks (defined by a given test) when the test is launched.')
    self.setMPILaunch(v)

    if self.useBatch == True:
      prompt = '[3] specify a constraint (e.g. "gpu" on Piz Daint) (optional - hit enter if not applicable):'
      v = py23input(prompt)
      self.setBatchConstraint(v)

      prompt = '[4] Account to charge (optional - hit enter if not applicable): '
      v = py23input(prompt)
      self.setHPCAccountName(v)

      prompt = '[5] Name of queue to submit tests to (optional - hit enter if not applicable): '
      v = py23input(prompt)
      self.setQueueName(v)

    self.writeDefinition()
    print('\n')
    print('** If you wish to change the config for your batch system, either')
    print('**  (i) delete the file',Launcher.confFileName,' or')
    print('** (ii) re-run with the command line arg --configure')
    print('----------------------------------------------------------------')

  def setup(self):
    try:
      self.loadDefinition()
    except PthTestHarnessLoadException :
      self.configure()
      self.writeDefinition()

  def writeDefinition(self):
    file = open(Launcher.confFileName,'w')
    major,minor,patch=getVersion()
    file.write('majorVersion=' + str(major) + '\n')
    file.write('minorVersion=' + str(minor) + '\n')
    file.write('patchVersion=' + str(patch) + '\n')
    file.write('queuingSystemType=' + self.queuingSystemType + '\n')
    file.write('mpiLaunch=' + self.mpiLaunch + '\n')
    if self.useBatch == True:
      file.write('accountName=' + self.accountName + '\n')
      file.write('batchConstraint=' + self.batchConstraint + '\n')
      file.write('queueName=' + self.queueName + '\n')
    file.close()

  def loadDefinition(self):
    try:
      majorFile = None
      minorFile = None
      patchFile = None
      file = open(Launcher.confFileName,'r')
      for v in file :
        key,value = v.split('=',1)
        value = value.rstrip()
        if key == 'majorVersion' :
          majorFile = int(value)
        if key == 'minorVersion' :
          minorFile = int(value)
        if key == 'patchVersion' :
          patchFile = int(value)
        if key == 'queuingSystemType' :
          self.setQueueSystemType(value)
        if key == 'mpiLaunch' :
          self.setMPILaunch(value)
        if self.useBatch == True:
          if key == 'batchConstraint' :
            self.setBatchConstraint(value)
          if key == 'queueName' :
            self.setQueueName(value)
          if key == 'accountName' :
            self.setHPCAccountName(value)
      file.close()
    except:
      raise PthTestHarnessLoadException('[pth] You must execute configure(), and or writeDefinition() first')

    # Do not accept conf files if the major.minor version is stale, or if versions are missing
    major,minor,patch = getVersion()
    if majorFile < major or (minorFile < minor and majorFile == major) or \
         majorFile==None or minorFile==None or patchFile==None :
      message = '[pth] Incompatible, outdated ' + Launcher.confFileName + ' file detected. Please delete it and re-run to reconfigure.'
      raise RuntimeError(message)

  def createSubmissionFile(self,testname,commnd,ranks,ranks_per_node,walltime,outfile):
    filename = ''
    if not self.useBatch:
      print('Warning: no submission file creation required')
      return(filename)

    if self.batchConstraint and self.queuingSystemType != 'slurm' :
      message = '[pth] Constraints are only currently supported with SLURM'
      raise RuntimeError(message)

    if self.queuingSystemType == 'pbs':
      filename = generateLaunch_PBS(self.accountName,self.queueName,testname,self.mpiLaunch,commnd,ranks,ranks_per_node,walltime,outfile)

    elif self.queuingSystemType == 'lsf':
      filename = generateLaunch_LSF(self.accountName,self.queueName,testname,self.mpiLaunch,commnd,ranks,None,walltime,outfile)

    elif self.queuingSystemType == 'slurm':
      filename = generateLaunch_SLURM(self.accountName,self.queueName,testname,self.batchConstraint,self.mpiLaunch,commnd,ranks,ranks_per_node,walltime,outfile)

    elif self.queuingSystemType == 'load_leveler':
      raise ValueError('[pth] Unsupported: LoadLeveler needs to be updated')

    print('Created submission file:',filename)
    return(filename)

  def submitJob(self,test):
    if test.use_sandbox:
      sandboxBack = os.getcwd()
      os.mkdir(test.sandbox_path) # error if  it already exists
      os.chdir(test.sandbox_path)
    test.setVerbosityLevel(self.verbosity_level)
    if not self.useBatch:
      mpiLaunch = self.mpiLaunch

      if self.mpiLaunch == 'none' and test.ranks != 1:
        print('[Failed to launch test \"' + test.name + '\" as test uses > 1 MPI ranks and no MPI launcher was provided]')
      else:
        if self.mpiLaunch == 'none':
          launchCmd = []
          for e in test.execute:
            launchCmd.append( e + " >> " + os.path.join(test.output_path,test.output_file) )
        else:
          launch = pthFormatMPILaunchCommand(mpiLaunch,test.ranks,None)
          launchCmd = []
          for e in test.execute:
            launchCmd.append( launch + ' ' + e + " >> " + os.path.join(test.output_path,test.output_file) )
        if self.verbosity_level > 0:
          if test.use_sandbox:
            for lc in launchCmd:
              print('[Executing from ' + os.getcwd() + ']',lc)
          else :
            for lc in launchCmd:
              print('[Executing]',lc)
        for lc in launchCmd:
          test.errno = os.system(lc) >> 8
    else:
      outfile = os.path.join(test.output_path,test.output_file)
      launchfile = self.createSubmissionFile(test.name,test.execute,test.ranks,'',test.walltime,outfile)
      launchCmd = self.jobSubmissionCommand + launchfile
      if self.verbosity_level > 0:
        if test.use_sandbox:
          print('[Executing from ' + os.getcwd() + ']',launchCmd)
        else :
          print('[Executing]',launchCmd)
      os.system(launchCmd)

    if test.use_sandbox:
        os.chdir(sandboxBack)

  def clean(self,test):
    print('[ -- Removing output for test:',test.name,'-- ]')
    if test.use_sandbox:
      sandboxBack = os.getcwd()
      if not os.path.isdir(test.sandbox_path) :
          os.mkdir(test.sandbox_path)
      os.chdir(test.sandbox_path)
    outfile = os.path.join(test.output_path,test.output_file)
    if os.path.isfile(outfile) :
      os.remove(outfile)
    if test.comparison_file != outfile and os.path.isfile(test.comparison_file) :
      foundInLocalTree = False
      cwd = os.getcwd()
      for (root, dirs, files) in os.walk(cwd) :
        for f in files :
          if os.path.abspath(test.comparison_file) == os.path.abspath(os.path.join(root,f)) :
            foundInLocalTree = True
            break
        if foundInLocalTree :
          break
      if foundInLocalTree :
        os.remove(test.comparison_file)
      else :
        message = "Refusing to remove output file " + test.comparison_file + " since it does not live in the local subtree. If you really wanted to compare with this file, please delete it yourself to proceed"
        raise RuntimeError(message)
    if self.useBatch:
      stderrFile = test.name + '.stderr'
      if os.path.isfile(stderrFile) :
        os.remove(stderrFile)
      stdoutFile = test.name + '.stdout'
      if os.path.isfile(stdoutFile) :
        os.remove(stdoutFile)
      if self.queuingSystemType == 'pbs':
        pbsFile = test.name + '-pth.pbs'
        if os.path.isfile(pbsFile) :
          os.remove(pbsFile)
      elif self.queuingSystemType == 'lsf':
        lsfFile = test.name + '-pth.lsf'
        if os.path.isfile(lsfFile) :
          os.remove(lsfFile)
      elif self.queuingSystemType == 'slurm':
        slurmFile = test.name + '-pth.slurm'
        if os.path.isfile(slurmFile) :
          os.remove(slurmFile)
      elif self.queuingSystemType == 'load_leveler':
        llqFile = test.name + '-pth.llq'
        if os.path.isfile(llqFile) :
          os.remove(llqFile)

    if test.use_sandbox:
      os.chdir(sandboxBack)
      shutil.rmtree(test.sandbox_path) # remove entire subtree

# Deprecated alias for backwards compatibility
pthLaunch = Launcher
