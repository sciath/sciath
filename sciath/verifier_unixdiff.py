
import os
from sciath.job import Job
from sciath.job import JobSequence
from sciath.job import JobDAG
from sciath.verifier import Verifier
from sciath import sciath_test_status
import sciath._subprocess as subp
from sciath.launcher import _removeFile


class VerifierUnixDiff(Verifier):
    def __init__(self,test,expected_file,output_file=None,**kwargs):
      Verifier.__init__(self,test,**kwargs)
      if expected_file is None:
          raise RuntimeError('[Sciath] Must set expected_file on test')
      self.expected_file = expected_file
      if output_file is None:
          self.output_file = self.o_name[-1]

    def execute(self,output_path,exec_path = None):
    
        self.status = None
        self.report = []
    
        if not os.path.isfile(self.expected_file) :
            self.report.append("[UnixDiff] Expected file \"" + self.expected_file + "\" was not found")
            self.status = sciath_test_status.expected_file_not_found
            return
   
        output_full_path = os.path.join(output_path,self.output_file)
        if not os.path.isfile(output_full_path) :
            self.report.append("[UnixDiff] Output file \"" + output_full_path + "\" was not found")
            self.status = sciath_test_status.output_file_not_found
            return
    
    
        stdoutfile = os.path.join(output_path,"sciath.verifier-unixdiff.stdout")
        file_o = open( stdoutfile, 'w')
        e = subp.run(['diff','-c',self.expected_file,self.output_file],file_o,None)
        file_o.close()
    
        if int(e) != 0:
            with open(stdoutfile, 'r') as f:
                data = f.readlines()
            self.report += data
            for k in range(0,len(self.report)):
                line = self.report[k]
                self.report[k] = line.rstrip("\n")
      
            self.report.append("[UnixDiff] Expected and output files are not identical")
            self.report.append("[UnixDiff] Expected file: \"" + self.expected_file + "\"")
            self.report.append("[UnixDiff] Output file  : \"" + self.output_file + "\"")
            self.status = sciath_test_status.not_ok
        else:
            self.status = sciath_test_status.ok
    
        _removeFile(stdoutfile)

        return


