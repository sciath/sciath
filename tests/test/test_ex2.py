#!/usr/bin/env python
import os

from sciath.job import Job
from sciath.job import JobSequence
from sciath.job import JobDAG
from sciath.launcher import Launcher
from sciath.test import Test
from sciath.verifier_unixdiff import VerifierUnixDiff

this_dir = os.path.dirname(os.path.realpath(__file__))

OUTPUT_PATH = os.path.join(os.getcwd(),'output')
VERBOSITY = 0
job_launcher = Launcher()
job_launcher.setVerbosityLevel(VERBOSITY)

def test_print(test, output_path):
    status,report = test.verifier.execute(output_path = output_path)
    rep = [test.name] + status
    print(rep)
    for l in report:
        print(l)

def test1_ud(): # result: pass
    cmd = ['sh' , os.path.join(this_dir,'write_test1_ud.sh') ]

    t = Test(Job(cmd, 'Test_1_ud'))
    t.verifier = VerifierUnixDiff(t, os.path.join(this_dir,"./expected/t1.expected"))

    job_launcher.submitJob( t.job, output_path = OUTPUT_PATH, exec_path = OUTPUT_PATH )
    t.verify(output_path = OUTPUT_PATH)
    test_print(t, output_path = OUTPUT_PATH)
    return t

def test2_ud(): # result: fail
    cmd = ['sh' , os.path.join(this_dir,'write_test2_ud.sh') ]

    t = Test(Job(cmd, 'Test_2_ud'))
    t.verifier = VerifierUnixDiff(t, os.path.join(this_dir,"./expected/t1.expected"))

    job_launcher.submitJob( t.job, output_path = OUTPUT_PATH, exec_path = OUTPUT_PATH )
    t.verify(output_path = OUTPUT_PATH)
    test_print(t, output_path = OUTPUT_PATH)
    return t


def main():
    try:
        os.mkdir(OUTPUT_PATH)
    except:
        pass

    # test using unix-diff verifierr
    t1_ud = test1_ud()
    t2_ud = test2_ud()

    # clean up
    clean_up = False
    if clean_up:
        for t in [t1_ud,t2_ud]:
            job_launcher.clean(t.job, output_path = OUTPUT_PATH)

main()
