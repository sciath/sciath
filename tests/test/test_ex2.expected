['Test_1_ud', 'pass', 'verification was successful']
['Test_2_ud', 'fail', 'verification failed']
--- <<TEST DIR STRIPPED>>/test/./expected/t1.expected
+++ <<TEST DIR STRIPPED>>/test_test2_sandbox/output/sciath.job-Test_2_ud.stdout
@@ -1,3 +1,2 @@
 aBc
 kspits=30
-
